### PR TITLE
[Tagger][PodWatcher] Detect kubernetes pod phase change

### PR DESCRIPTION
**Pull request by nammn**
### What does this PR do?

- pods that exit (e.g. jobs) finish and change their state from `running` to `succeed` or `fail` 

The tagger tries to only update tags for pods where we have calculated changes. Currently detections are only for `metadata`, `container` and `staticpods`.
Pods that have exited after a long run can be missed by the tagger. Therefore this PR watches the `phase` of the pod as well.

### Motivation

- missmatch between tag:`pod_phase` and actual `phase`
- jira: CAP-95

### Additional Notes

It does touch the podWatcher to make sure we are safe I added some additional tests and put it under the isWatchingTags part.

### Describe your test plan

Write there any instructions and details you may have to test your PR.

### CODE REVIEW

File: releasenotes/notes/fix-podphase-tagging-2db951956eeed605.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+# CHANGELOG.rst is rendered. So the text needs to be worded so that
+# it does not depend on any information only available in another
+# section. This may mean repeating some details, but each section
+# must be readable independently of the other.
+#
+# Each section note must be formatted as reStructuredText.
+---
+fixes:
+  - |
+    Fix a bug when a kubernetes job has exited after some time the tagger does not update it even if it did change it's state.
```
Comment (hkaj) [inline releasenotes/notes/fix-podphase-tagging-2db951956eeed605.yaml:11]:
```suggestion
    Fix a bug when a kubernetes job has exited after some time the tagger does not update it even if it did change its state.
```

**Review by hkaj — APPROVED** [review]
Looks great! Left a grammar fix, otherwise feel free to merge.

File: pkg/util/kubernetes/kubelet/podwatcher.go [hunk 2]
```diff
@@ -123,16 +128,20 @@ func (w *PodWatcher) computeChanges(podList []*Pod) ([]*Pod, error) {
 
 		// Detect changes in labels and annotations values
 		newLabelsOrAnnotations := false
+		newPhase := false
```
Comment (ahmed-mez) [inline pkg/util/kubernetes/kubelet/podwatcher.go:131]: nit: could you update the comment in L129 accordingly as we detect phase changes now?

File: pkg/util/kubernetes/kubelet/podwatcher.go [hunk 3]
```diff
@@ -65,7 +67,9 @@ func (w *PodWatcher) PullChanges() ([]*Pod, error) {
 	if err != nil {
 		return podList, err
 	}
-	return w.computeChanges(podList)
+	changes, err := w.computeChanges(podList)
+
+	return changes, err
```
Comment (ahmed-mez) [inline pkg/util/kubernetes/kubelet/podwatcher.go:72]: out of curiosity, what's the rationale behind this change?

Comment (nammn) [reply pkg/util/kubernetes/kubelet/podwatcher.go:72]: Oh, this is a leftover from my debugging. Thanks for the catch. Changing it 👍

File: releasenotes/notes/fix-podphase-tagging-2db951956eeed605.yaml (hunk 1 above)
Comment (nammn) [reply releasenotes/notes/fix-podphase-tagging-2db951956eeed605.yaml]: added it, thanks!

File: pkg/util/kubernetes/kubelet/podwatcher.go (hunk 2 above)
Comment (nammn) [reply pkg/util/kubernetes/kubelet/podwatcher.go:130]: Done, makes sense!

**Review by ahmed-mez — APPROVED** [review]
Left few question but looks good overall, thanks for the fix 🙇 
Even though the bug was discovered by the process agent + job tags as described in the release note, it also affects all tagger clients (not only process agent) and all k8s resources (not only jobs). Could you update the PR title accordingly? (suggestion: `[Tagger][PodWatcher] Detect kubernetes pod phase change`)
thanks!
